### PR TITLE
[SYCL][L0] Add SubDevices into the device cache

### PR DESCRIPTION
**Pull request by bso-intel**
This PR adds the missing sub-devices in the device cache.
This fixes the missing cache loopup for sub-devices.

Signed-off-by: Byoungro So <byoungro.so@intel.com>

### CODE REVIEW

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 1]
```diff
@@ -1911,29 +1914,25 @@ pi_result piextDeviceCreateWithNativeHandle(pi_native_handle NativeHandle,
...
-  std::lock_guard<std::mutex> Lock(Platform->PiDevicesCacheMutex);
-  pi_result Res = populateDeviceCacheIfNeeded(Platform);
-  if (Res != PI_SUCCESS) {
-    return Res;
+  {
+    std::lock_guard<std::mutex> Lock(Platform->PiDevicesCacheMutex);
+    pi_result Res = populateDeviceCacheIfNeeded(Platform);
+    if (Res != PI_SUCCESS) {
+      return Res;
+    }
```
Comment (bso-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:1922]: Just added braces { } to remove the lock_guard before we call Platform->getDeviceFromNativeHandle() below, which also will use the lock_guard in there.

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 2]
```diff
@@ -1825,18 +1825,21 @@ pi_result piDevicePartition(pi_device Device,
...
-      pi_result Result = OutDevices[I]->initialize();
+      std::unique_ptr<_pi_device> PiSubDevice(
+          new _pi_device(ZeSubdevices[I], Platform));
+      pi_result Result = PiSubDevice->initialize();
       if (Result != PI_SUCCESS) {
         delete[] ZeSubdevices;
         return Result;
       }
+      OutDevices[I] = PiSubDevice.get();
+      Platform->PiDevicesCache.push_back(std::move(PiSubDevice));
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:1842]: You are going to add "new" sub-devices every time piDevicePartition is called, but the idea is that this should only be done once (analogous to how we add root-devices only once to the cache). NOTE: every call to zeDeviceGetSubDevices
 will return the same Level Zero handles.

**Review by smaslov-intel — CHANGES_REQUESTED** [review]

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 2 above)
Comment (bso-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:1842]: Yes, the current piDevicePartition already behaved like that.
We already created a new _pi_device objects for all returned sub-devices from zeDeviceGetSubDevices().
I was only adding the caching capability to this existing behavior.
Do you want to change the current behavior to check the cache before we call zeDeviceGetSubDevices() in addition to caching?
If so, please advise me how to handle multiple levels of partitioning.
AFAIK, a device can be partitioned into 2 subdevices first and then another level of partitioning will partition them into 4 subdevices. 
How can we handle them? Any hint would be appreciated.

Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:1842]: > AFAIK, a device can be partitioned into 2 subdevices first and then another level of partitioning will partition them into 4 subdevices.

Let's not handle this yet (not yet truly supported scenario). You can just add an assertion that the input device is a root-device.

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 3]
```diff
@@ -1802,19 +1802,24 @@ pi_result piDevicePartition(pi_device Device,
 
   PI_ASSERT(Device, PI_INVALID_DEVICE);
 
+  // Check if Device was already partitioned into the same or bigger size
+  // before. If so, we can return immediately without searching the global
+  // device cache. Note that L0 driver always return the same handles in the
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:1807]:
```suggestion
  // device cache. Note that L0 driver always returns the same handles in the
```

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 4]
```diff
@@ -1825,17 +1830,27 @@ pi_result piDevicePartition(pi_device Device,
...
-      pi_result Result = OutDevices[I]->initialize();
-      if (Result != PI_SUCCESS) {
-        delete[] ZeSubdevices;
-        return Result;
+      pi_device Dev = Platform->getDeviceFromNativeHandle(ZeSubdevices[I]);
+      if (Dev) {
+        OutDevices[I] = Dev;
+      } else {
+        std::unique_ptr<_pi_device> PiSubDevice(
+            new _pi_device(ZeSubdevices[I], Platform));
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:1844]: I understand this is not a new problem, but who will delete these? I think you should call piDeviceRelease of sub-devices in piDeviceRelease of root-device

File: sycl/plugins/level_zero/pi_level_zero.hpp [hunk 5]
```diff
@@ -156,6 +156,9 @@ struct _pi_device : _pi_object {
   // Level Zero device handle.
   ze_device_handle_t ZeDevice;
 
+  // Keep the subdevices that are partitioned from this pi_device for reuse
+  std::vector<pi_device> SubDevices;
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.hpp:160]: Please add a note that the order is significant and repeats the order in which L0 returns handles

**Review by smaslov-intel — CHANGES_REQUESTED** [review]

File: sycl/plugins/level_zero/pi_level_zero.hpp (hunk 5 above)
Comment (bso-intel) [reply sycl/plugins/level_zero/pi_level_zero.hpp:163]: Thanks. Done.

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 4 above)
Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:1844]: Please check/add

Comment (bso-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:1844]: Yes, I verified with valgrind, and they are released properly. No memory leak is detected for sub-devices.
Anyway, if SYCL RT does not call piDeviceRelease for pi_device returned from the plugin, it is a bug in SYCL RT.

Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:1844]: But if SYCL RT **does** call `piDeviceRelease ` on every PI device returned from `piDevicePartition` then your changes will break it: multiple partitioning calls will now return same PI devices, and so they will be deleted too early, at the first `piDeviceRelease` of a sub-device.

Comment (bso-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:1844]: I see. 
Then, I will call piDeviceRetain() when we reuse the same pi_device for the same sub-device.

Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:1844]: Yes, and add a test covering this.

Comment (bso-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:1846]: okay. I will extend the existing test https://github.com/intel/llvm-test-suite/blob/intel/SYCL/Basic/query.cpp to test reusing the same sub-device.
It is going to be a separate PR in that repo later.

**Review by smaslov-intel — APPROVED** [review]
LGTM, remember to add test

**Comment by bso-intel** [comment]
It seems the failure FAIL: Clang::intel-fpga-global-const.cpp is not related to my changeset.
I cannot reproduce the issue locally.
Some other PRs also show the same issue.

**Review by smaslov-intel — APPROVED** [review]
LGTM

**Comment by bso-intel** [comment]
I tried to rerun 5 times today, and FAIL: Clang::intel-fpga-global-const.cpp is persistent today.
I am pretty sure this failure is not related to my changset.
I cannot reproduce the failure on my local machine, this failure started from the last 2 commits which are limited to Level Zero, but it complains that -fsycl is unknown argument.

@bader 
I looked at other PR like https://github.com/intel/llvm/pull/3315 also suffers from this particular test case failure.
Could you help resolve this issue?

**Comment by bader** [comment]
It looks like we have a conflicts between https://github.com/intel/llvm/commit/a539197e461aaf9f1c5f3a7be7e3ee32bea207a9 and https://github.com/intel/llvm/commit/f32c431d57a2a340e814ec51f7e132293432a576. I'll try to fix this issue..
@smanna12, @AaronBallman, FYI.

**Comment by bader** [comment]
I hope https://github.com/intel/llvm/pull/3339 addresses the problem with clang/test/SemaSYCL/intel-fpga-global-const.cpp.

**Comment by bso-intel** [comment]
Thanks a lot, @bader .
I am waiting for your PR to get resolved.

**Comment by bader** [comment]
I'll restart buildbot tests in this PR after #3339 is merged.

**Comment by bso-intel** [comment]
Great! Then, I can go to bed. Thank you again, Alexey.

**Pull request merged**